### PR TITLE
[Landcover Toolkit] Treat Datasets as immutable

### PR DESCRIPTION
**Pull request by gino-m**
This allows references to Dataset, Landsat8, etc. to behave as we would expect; like EE API, objects are considered immutable, and mutators return a new instance, allowing us to do things like this:
```js
var l8 = lct.Landsat8();
var cloudFreeL8 = l8.fmaskCloudsAndShadows();
Map.addLayer(l8, l8.getDefaultVisParams(), 'Raw Landsat8');
Map.addLayer(cloudFreeL8, cloudFreeL8.getDefaultVisParams(), 'Cloud-free Landsat8');
```

### CODE REVIEW

File: toolkits/landcover/impl/dataset.js [hunk 1]
```diff
@@ -48,8 +61,9 @@ var Dataset = function(id, defaultVisParams) {
 Dataset.prototype.filterDate = function(start, end) {
   // TODO(gino-m): Implement month and year ranges.
   // TODO(gino-m): Implement single day/month/year.
-  this.collection_ = this.collection_.filterDate(start, end);
-  return this;
+  var dataset = this.clone_();
```
Comment (gorelick-google) [inline toolkits/landcover/impl/dataset.js]: This is gross.  Are there other options?

File: toolkits/landcover/impl/dataset.js [hunk 2]
```diff
@@ -153,33 +168,34 @@ Dataset.prototype.computeCommonBandNames_ = function(bandNames) {
  * @return {!Dataset}
  */
 Dataset.prototype.addBandIndices = function(var_names) {
+  var dataset = this.clone_();
   var indices = Array.prototype.slice.call(arguments);
 
   // Compute the transformation to common band names.
-  var bandNames = this.collection_.first().bandNames();
+  var bandNames = dataset.collection_.first().bandNames();
```
Comment (gorelick-google) [inline toolkits/landcover/impl/dataset.js]: There's an unclean mixture of uses of this and dataset in here.
Once you've cloned, you shouldn't use this any more, so either
use dataset throughout or don't clone until you're about to modify state (#185)

File: toolkits/landcover/impl/landsat8.js [hunk 3]
```diff
@@ -60,19 +64,27 @@ Landsat8.prototype = Object.create(Dataset.prototype);
 Landsat8.prototype.COMMON_BAND_NAMES = COMMON_BAND_NAMES;
 
 /**
- * Masks clouds and shadows using relevant bits in the `pixel_qa` band generated
- * by the CFMASK algorithm.
+ * Returns a copy of this dataset, copying all related state. Used internally
```
Comment (gorelick-google) [inline toolkits/landcover/impl/landsat8.js]: If there's no extra state, is there a reason to override?

File: toolkits/landcover/impl/dataset.js (hunk 1 above)
Comment (gino-m) [reply toolkits/landcover/impl/dataset.js]: Yes, it's icky. A few other options come to mind:

1. Don't make Dataset immutable, and continue treating Datasets as mutable [Builders](https://medium.com/@axelhadfeg/builder-pattern-using-javascript-and-es6-ec1539182e24) (link is for ES6, but same idea applies)
2. Add all state/fields to constructor and call here (worse).
3. Implement a util to clone the class dynamically (messy).

Any preference?

File: toolkits/landcover/impl/dataset.js (hunk 2 above)
Comment (gino-m) [reply toolkits/landcover/impl/dataset.js]: The two instances are identical at this point, so I'd say the one reference to `this` in the following line is inconsistent, but "unclean" is a bit of a stretch. I'll update to make them consistent.

File: toolkits/landcover/impl/landsat8.js (hunk 3 above)
Comment (gino-m) [reply toolkits/landcover/impl/landsat8.js]: If we enable type checking using our JSDoc we need to override this to create and return an instance of the right class.

File: toolkits/landcover/impl/landsat8.js [hunk 4]
```diff
@@ -41,16 +42,19 @@ var QA_BAND = 'pixel_qa';
  * Returns a new Landsat 8 SR dataset instance.
  *
  * @constructor
+ * @param {ee.ImageCollection=} collection The Landsat image collection backing
+ *    this dataset. If not specified, Landsat 8 SR will be used.
  * @return {!Landsat8}
  */
-var Landsat8 = function() {
+var Landsat8 = function(collection) {
```
Comment (gino-m) [inline toolkits/landcover/impl/landsat8.js]: If we treat datasets as builders this arg can go away along with the clone() operations. Wdyt?

**Comment by gino-m** [comment]
Abandoning this PR for now until we decide we actually want these classes to be immutable.

**Pull request closed without merging**